### PR TITLE
Augmented literal strings

### PR DESCRIPTION
**Pull request by R4PaSs**
As a follow-up to #1991, here's the support mentioned in #1734 for bytestrings, regex and raw strings

NOTE: Depends on #1991 for integration, you may review the last 3 commits while #1991 is reviewed and eventually merged

TODO: Support degraded mode (Bytes and Byte only) for byte SuperStrings instead of refusing at compile-time

### CODE REVIEW

File: src/compiler/abstract_compiler.nit [hunk 1]
```diff
@@ -3575,13 +3587,63 @@ redef class AArrayExpr
...
+	redef fun expr(v) do
+		var s: RuntimeVariable
+		if is_string then
+			s = v.string_instance(value)
+		else if is_bytestring then
+			var ns = v.native_string_instance(bytes.items, bytes.length)
+			var ln = v.int_instance(bytes.length)
+			var res = v.send(v.get_property("to_bytes_with_copy", ns.mtype), [ns, ln])
+			if res == null then
+				print "Cannot call property `to_bytes` on {self}"
```
Comment (privat) [inline src/compiler/abstract_compiler.nit]: wat? the print seems bad at this point

File: src/rapid_type_analysis.nit [hunk 2]
```diff
@@ -223,6 +223,8 @@ class RapidTypeAnalysis
 		force_alive("UInt16")
 		force_alive("Int32")
 		force_alive("UInt32")
+		force_alive("Bytes")
+		force_alive("Regex")
```
Comment (privat) [inline src/rapid_type_analysis.nit]: please remove forcealive

File: src/rapid_type_analysis.nit [hunk 3]
```diff
@@ -585,6 +587,12 @@ redef class AStringFormExpr
 		v.add_type(native)
 		var prop = v.get_method(native, "to_s_full")
 		v.add_monomorphic_send(native, prop)
+		for i in services do v.add_callsite(i)
+		if is_bytestring then
+			var nstype = v.analysis.mainmodule.native_string_type
+			var meth = v.get_method(nstype, "to_bytes_with_copy")
```
Comment (privat) [inline src/rapid_type_analysis.nit]: move in typing

File: src/semantize/typing.nit [hunk 4]
```diff
@@ -1469,19 +1469,79 @@ redef class ACharExpr
...
+		end
 		if mclass == null then return # Forward error
 		self.mtype = mclass.mclass_type
 	end
 end
 
+redef class AStringFormExpr
+
+	# Added services because of prefixes
+	var services = new Array[CallSite]
```
Comment (privat) [inline src/semantize/typing.nit]: one attribute per exact service so you can use them in analysis and engines

File: src/compiler/abstract_compiler.nit [hunk 5]
```diff
@@ -3632,9 +3698,39 @@ redef class ASuperstringExpr
...
+				end
+				if i == 'b' then
+					var prop = v.get_property("extended=", res.mtype)
+					v.send(prop, [res, v.bool_instance(false)])
+					continue
+				end
+				# Should not happen, this needs to be updated
+				# along with the addition of new suffixes
+				abort
+			end
```
Comment (xymus) [inline src/compiler/abstract_compiler.nit]: That's a big duplicated block of code here... is there a difference to the implementation on AStringExpr?

File: src/interpreter/naive_interpreter.nit [hunk 6]
```diff
@@ -1980,6 +2038,33 @@ redef class ASuperstringExpr
...
+				end
+				if j == 'b' then
+					var prop = v.force_get_primitive_method("extended=", res.mtype)
+					v.send(prop, [res, v.bool_instance(false)])
+					continue
+				end
+				# Should not happen, this needs to be updated
+				# along with the addition of new suffixes
+				abort
+			end
```
Comment (xymus) [inline src/interpreter/naive_interpreter.nit]: In the interpreter too this block is duplicated. This big if/else could probably be moved up to a common super class of AStringExpr and ASuperstringExpr.

**Comment by R4PaSs** [comment]
Fixed most of the remarks @privat and @xymus made.

File: src/compiler/abstract_compiler.nit [hunk 7]
```diff
@@ -3574,14 +3586,68 @@ redef class AArrayExpr
 	end
 end
 
+redef class AugmentedStringFormExpr
+	# Factorize the making of a `Regex` object from a literal prefixed string
+	protected fun make_re(v: AbstractCompilerVisitor, rs: RuntimeVariable): nullable RuntimeVariable do
+		var res = v.send(v.get_property("to_re", rs.mtype), [rs])
```
Comment (privat) [inline src/compiler/abstract_compiler.nit]: use `self.to_re` that was identified by typing

File: src/compiler/abstract_compiler.nit [hunk 8]
```diff
@@ -3574,14 +3586,68 @@ redef class AArrayExpr
...
+		if res == null then
+			print "Cannot call property `to_re` on {self}"
+			abort
+		end
+		for i in suffix.chars do
+			if i == 'i' then
+				var prop = v.get_property("ignore_case=", res.mtype)
```
Comment (privat) [inline src/compiler/abstract_compiler.nit]: same thing for all the properties identified in typing

File: src/compiler/abstract_compiler.nit [hunk 9]
```diff
@@ -3574,14 +3586,68 @@ redef class AArrayExpr
...
+end
+
+redef class AStringExpr
+	redef fun expr(v) do
+		var s = v.string_instance(value)
+		if is_string then return s
+		if is_bytestring then
+			var ns = v.native_string_instance(bytes.items, bytes.length)
+			var ln = v.int_instance(bytes.length)
+			var res = v.send(v.get_property("to_bytes_with_copy", ns.mtype), [ns, ln])
```
Comment (privat) [inline src/compiler/abstract_compiler.nit]: to_bytes_with_copy seems also to be identified in typing

File: src/interpreter/naive_interpreter.nit [hunk 10]
```diff
@@ -1962,11 +1972,64 @@ redef class AArrayExpr
 	end
 end
 
+redef class AugmentedStringFormExpr
+	# Factorize the making of a `Regex` object from a literal prefixed string
+	fun make_re(v: NaiveInterpreter, rs: Instance): nullable Instance do
+		var res = v.send(v.force_get_primitive_method("to_re", rs.mtype), [rs])
```
Comment (privat) [inline src/interpreter/naive_interpreter.nit]: like in the compiler, reuse the methods already identified

**Comment by privat** [comment]
seems good modulo the last nitpicks

**Comment by R4PaSs** [comment]
Reroll: last nitpicks integrated

**Pull request merged**